src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,479 @@
+        else {
+            /*the number of vertices is defined before. To know the number of vertices ahead it would require
+            to run the loop once before, which would double computation time.
+            If vertices are not filled then there're visual bugs so there are here filled with dummy values.*/
+            vertex_coords[2 * i].setValue(0, 0, 0);
+            vertex_coords[2 * i + 1].setValue(0, 0, 0);
+        }
+    }
+
+    // horizontal lines
+    int i_offset_y = static_cast<int>(minY / Step) - vlines;
+    for (int i = vlines; i < nlines; i++) {
+        int iStep = (i + i_offset_y);
+        if (((iStep % numberSubdiv == 0) && divLines) || ((iStep % numberSubdiv != 0) && subDivLines)) {
+            vertex_coords[2 * i].setValue(minX, iStep * Step, 0);
+            vertex_coords[2 * i + 1].setValue(maxX, iStep * Step, 0);
+        }
+        else {
+            vertex_coords[2 * i].setValue(0, 0, 0);
+            vertex_coords[2 * i + 1].setValue(0, 0, 0);
+        }
+    }
+    vts->vertex.finishEditing();
+
+    parent->addChild(vts);
+    parent->addChild(grid);
+
+    return GridRoot;
+}
+
+void GridExtensionP::setEnabled(bool enable)
+{
+    enabled=enable;
+
+    drawGrid();
+
+}
+
+bool GridExtensionP::getEnabled()
+{
+    return enabled;
+}
+
+void GridExtensionP::createEditModeInventorNodes()
+{
+    // Create Grid Coin nodes ++++++++++++++++++++++++++++++++++++++++++
+    GridRoot = new SoSeparator();
+    GridRoot->ref();
+    GridRoot->setName("GridRoot");
+
+}
+
+SoSeparator * GridExtensionP::getGridRoot()
+{
+    return GridRoot;
+}
+
+void GridExtensionP::drawGrid(bool cameraUpdate) {
+    if (vp->ShowGrid.getValue() && enabled) {
+        createGrid(cameraUpdate);
+    }
+    else {
+        Gui::coinRemoveAllChildren(GridRoot);
+    }
+}
+
+void GridExtensionP::OnChange(Base::Subject<const char*> &rCaller, const char * sReason)
+{
+    (void) rCaller;
+
+    if (strcmp(sReason, "UserSchema") == 0) {
+        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+        unitsUserSchema = hGrp->GetInt("UserSchema", 0); //2 3 5 7 are imperial schemas. 2 3 inches, 5 7 feet
+    }
+}
+
+const char* ViewProviderGridExtension::GridStyleEnums[] = { "Dashed","Light",nullptr };
+App::PropertyQuantityConstraint::Constraints ViewProviderGridExtension::GridSizeRange = { 0.001,DBL_MAX,1.0 };
+
+ViewProviderGridExtension::ViewProviderGridExtension()
+{
+
+    EXTENSION_ADD_PROPERTY_TYPE(ShowGrid, (false), "Grid", (App::PropertyType)(App::Prop_None), "Switch the grid on/off");
+    EXTENSION_ADD_PROPERTY_TYPE(GridSize, (10.0), "Grid", (App::PropertyType)(App::Prop_None), "Gap size of the grid");
+    EXTENSION_ADD_PROPERTY_TYPE(GridStyle, (0L), "Grid", (App::PropertyType)(App::Prop_None), "Appearance style of the grid");
+    EXTENSION_ADD_PROPERTY_TYPE(GridSnap, (false), "Grid", (App::PropertyType)(App::Prop_None), "Switch the grid snap on/off");
```

Comment:
My proposal would be :
 * Remove `GridSnap` property from `ViewProviderGridExtension`
 * Offer utility functions in `ViewProviderGridExtension` that can be used by other classes (for example `VPSketch` -- or `SnapManager` soon -- to implement snapping

For example something like:
```
// This function updates 'x' and 'y' to closest grid point and returns how far original point is from the grid (in % of grid size)
float getClosestGridPoint(float &x, float &y) const;
```

Notice it isn't a very deep thought. Just an idea. Maybe not good. :)

Reply:
With respect to the utility functions I think it is not immediately useful for the current grid, but it can be very useful if we consider how the grid may be extended in the future.

Today we have a cartesian grid. If we assume we will always have one cartesian grid and one alone, then it may not be that much appealing to have that function. I mean we could have it, but it is not that important.

If we think that in the future we could provide other types of grid, such as a polar grid, or even a custom grid following a transformation from one space to another. Then, the situation drastically changes and this kind of function becomes key in separating grid definition/representation from (grid) snapping.

Talking aloud, we could even have a ViewProvider, where upon requiring an angle snap, the Grid switches to polar with the origin in the reference point, allowing to snap not only in angle, but within a (multiple of a) distance (as defined by the grid). This I do not know if it is useful or not, but a framework open to extension without modification looks to me like a good framework.

So yes, I think we should define such interface, which means: (a) modifying this class and  (b) using this function in the Snap functionality. But I propose to make this effort part of another PR.

@0penBrain 
@PaddleStroke 

I have been thinking about the Snap PR and this one and GridSnap.

Options:
A) GridSnap should never had become a property in the first place (not even of VPSketch). We remove it altogether. In Snap PR we substitute the GridSnap by some "Enabled" boolean controlled by some control of snapping. The Sketch will not remember if it was created using which snapping mode. Snapping depends on what is currently selected as snap mode. As we remove GridSnap, we do need some control to enable the user to manage snap mode (select which snapping mode should be used, as with ctrl alone this is not possible).

B) GridSnap is a edition property of individual sketches (i.e. VP of the sketch). The user expects that a sketch remembers it and next time it enters edit mode, it applies it. In view of Snap PR, the property should change to a Enum property. Given that Snap is provided by VPSketch (via delegation), a new property SnapMode should be created. A UI control modifies the property and the Snap PR takes this configuration into account to decide on snapping. Ctrl can still be used as a modifier, and may even offer different behaviour depending on the selection in the UI.

I think we do need a decision on this now, potentially impacting both PRs.

We should first discuss the behaviour we want to implement, so that we select an implementation that can support what we want to do (and preferably is open to extension in the future).

Let's go a round of short discussion. What is your opinion on this?

Reply:
Option A. Being grid snapping enabled or not shall just be a user setting that is valid application wide. :)

Reply:
Thank you. I am happy with that option. I can cross-out that open topic and bring the changes.